source/lac/dynamic_sparsity_pattern.cc

```diff
@@ -554,6 +554,47 @@ DynamicSparsityPattern::n_nonzero_elements() const
 }
 
 
+
+IndexSet
+DynamicSparsityPattern::nonempty_cols() const
+{
+  std::set<types::global_dof_index> cols;
+  for (const auto &line : lines)
+    cols.insert(line.entries.begin(), line.entries.end());
+
+  IndexSet res(this->n_cols());
+  res.add_indices(cols.begin(), cols.end());
+  return res;
+}
+
+
+
+IndexSet
+DynamicSparsityPattern::nonempty_rows() const
+{
+  const IndexSet  all_rows = complete_index_set(this->n_rows());
+  const IndexSet &rowset_  = rowset.size() == 0 ? all_rows : rowset;
```

Comment:
I'm not fond of the underscore idea. How about name this `locally_stored_rows`?

Reply:
> I'm not fond of the underscore idea.

the idea is to make it similar to member variable `rowset`.

Reply:
nevermind, I will rename